### ISSUE TITLE
Support additional types in HTMLRewriter

### ISSUE THREAD

**Issue by Cherry** [cloudflare/workerd#1508]
Since bun has now decided to add support for passing strings and ArrayBuffers to `HTMLRewriter .transform`, it would probably be a good idea to follow suit and implement that here. too.

https://twitter.com/jarredsumner/status/1738891979323580756

**Comment by Cherry** [comment]
References:

- https://github.com/oven-sh/bun/issues/2325
- https://github.com/oven-sh/bun/pull/7797

**Comment by jasnell** [comment]
It's worth evaluating options here but we won't do this simply to chase whatever changes bun decides to unilaterally make to the API. I definitely think we would be open to standardizing the `HTMLRewriter` API (perhaps within WinterCG) which would hopefully encourage collaboration instead of unilateral changes.

In this specific case, bun's changes are a bit problematic. The `transform(...)` API was originally designed to allow the transformation itself to be performed asynchronously, and `workerd`'s implementation does exactly that. Even tho the `Response` object is returned immediately, the underlying transformation happens async and the entire internal implementation is built around that assumption.

Unfortunately, however, bun decided to implement this piece entirely synchronously and if a `string` or `ArrayBuffer` is passed in, the assumption is that you'll synchronously get the transformed `string` or `ArrayBuffer` back out. That's not a trivial change to make and I'm not nearly convinced there's value in actually putting in the effort to make that change.

What I think would have been a much better alternative is adding an alternative API alongside `transform(...)` that introduced the new semantics rather than modifying `transform(...)` itself, and having that new API return a `Promise<T>` rather than `T`:

```
await rewriter.transformValue('<h3>hi</h3>');
await rewriter.transformValue(arrayBuffer);
```

This is an approach I think we could absolutely implement pretty easily.

Going to keep evaluating this, however. For now, however, no work on this is planned.

**Comment by jasnell** [comment]
Just documenting to inform any future work we might do here:

With bun's changes... 

* if `string` is passed in to `transform(...)`, a `string` is returned.
* If an `ArrayBuffer` *or any `ArrayBufferView` (e.g. `Uint8Array`, `Uint32Array`, `Float32Array`, etc) is passed in, an `ArrayBuffer` is returned.
* There is no indication which character encoding is assumed for `ArrayBuffer`/`ArrayBufferView` input, but I assume it's UTF-8. We would need to get clarification on that. (note that we also do not specify this in the CF implementation)
* Should the `string`/`ArrayBuffer` input ultimately be treated as `UsvString` or `ByteString`? Based on testing it appears `ByteString` but we should verify.

**Comment by kentonv** [comment]
What does Bun do if you ask for a synchronous transformation, but you've registered asynchronous callbacks?

OK, James and I tested it. It appears that `transform()` will _synchronously wait_ for async callbacks to complete. In other words, `transform()` is a synchronous function that can recursively pump the event loop. Unrelated microtasks may run before `transform()` returns.

Of course, this implies that any old synchronous code in Bun can synchronously wait on promises (i.e. without being declared async itself, or using `.then()`). That's quite a major change to the JavaScript async model! I suppose you could argue that it's a feature, not a bug -- it "solves" the old "function coloring" problem. But it's also has deep implications for application code. Where previously you could assume that two async continuations could not interleave, now there is no such guarantee.

Needless to say, we don't plan to allow synchronously awaiting promises in the workers runtime.

**Comment by jasnell** [comment]
To illustrate by example, 

This works currently in the Workers implementation... but hangs indefinitely in Bun.

```
let n = 0;

let r;
let p = new Promise((res) => r = res);

let m = new HTMLRewriter().on('h3', {
  async text(content) {
    console.log(n++, 'a');
    await p;
    console.log(n++, 'b');
    content.replace('hello');
    console.log(n++, 'c');
  }
}).transform(new Response('<h3>hi</h3>'));

console.log(n++, 'd');
r();

console.log(await m.text());
```

**Comment by kentonv** [comment]
Jarred says on Twitter that this is a bug: async callbacks should not be allowed when performing synchronous transforms.

https://twitter.com/jarredsumner/status/1742368420199588208

OK, so assuming synchronous `transform()` would instead throw an exception if any callback returned a Promise, I guess it's not an unreasonable feature, and lol-html itself could support this approach, but we'd have to make big changes to our driver code which currently is designed to be async.

**Comment by harrishancock** [comment]
> (note that we also do not specify [character encoding] in the CF implementation)

@jasnell do you mean when calling `rewriter.transform(response)`? As I recall, the need to know the encoding is, in fact, the primary reason that `transform()` currently only accepts Response objects -- we read it from the Content-Type: https://github.com/cloudflare/workerd/blob/main/src/workerd/api/html-rewriter.c%2B%2B#L1042-L1054

**Comment by jasnell** [comment]
Sorry, yeah, I wasn't clear at all. I meant that the API itself does not clarify how the encoding is handled. The docs (https://developers.cloudflare.com/workers/runtime-apis/html-rewriter/) do not mention charsets, encodings, or the `Content-Type` header at all.
